src/spec-configuration/httpOCIRegistry.ts

```diff
@@ -0,0 +1,233 @@
+			const serviceGroup = serviceRegex.exec(wwwAuthenticate);
+			const scopeGroup = scopeRegex.exec(wwwAuthenticate);
+
+			if (!realmGroup || !serviceGroup || !scopeGroup) {
+				output.write(`[httpOci] WWW-Authenticate header is not in expected format. Got:  ${wwwAuthenticate}`, LogLevel.Trace);
+				return;
+			}
+
+			const wwwAuthenticateData = {
+				realm: realmGroup[1],
+				service: serviceGroup[1],
+				scope: scopeGroup[1],
+			};
+
+			const bearerToken = await fetchRegistryBearerToken(params, ociRef, wwwAuthenticateData);
+			if (!bearerToken) {
+				output.write(`[httpOci] ERR: Failed to fetch Bearer token from registry.`, LogLevel.Error);
+				return;
+			}
+
+			httpOptions.headers.authorization = `Bearer ${bearerToken}`;
+			break;
+
+		default:
+			output.write(`[httpOci] ERR: Unsupported authentication mode '${wwwAuthenticate.split(' ')[0]}'`, LogLevel.Error);
+			return;
+	}
+
+	// Retry the request with the updated authorization header.
+	const reattemptRes = await requestResolveHeaders(httpOptions, output);
+	output.write(`[httpOci] ${reattemptRes.statusCode} on reattempt after auth: ${httpOptions.url}`, LogLevel.Trace);
+
+	// Cache the auth header if the request did not result in an unauthorized response.
+	if (reattemptRes.statusCode !== 401) {
+		params.cachedAuthHeader = httpOptions.headers.authorization;
+	}
+
+	return reattemptRes;
+}
+
+// Attempts to get the Basic auth credentials for the provided registry.
+// These may be programatically crafted via environment variables (GITHUB_TOKEN),
+// parsed out of a special DEVCONTAINERS_OCI_AUTH environment variable,
+async function getBasicAuthCredential(params: CommonParams, ociRef: OCIRef | OCICollectionRef): Promise<string | undefined> {
+	const { output, env } = params;
+	const { registry } = ociRef;
+
+	// TODO: Ask docker credential helper for credentials.
+
+	if (!!env['GITHUB_TOKEN'] && registry === 'ghcr.io') {
+		output.write('[httpOci] Using environment GITHUB_TOKEN for auth', LogLevel.Trace);
+		const userToken = `USERNAME:${env['GITHUB_TOKEN']}`;
+		return Buffer.from(userToken).toString('base64');
+	} else if (!!env['DEVCONTAINERS_OCI_AUTH']) {
+		// eg: DEVCONTAINERS_OCI_AUTH=service1|user1|token1,service2|user2|token2
+		const authContexts = env['DEVCONTAINERS_OCI_AUTH'].split(',');
+		const authContext = authContexts.find(a => a.split('|')[0] === registry);
+
+		if (authContext) {
+			output.write(`[httpOci] Using match from DEVCONTAINERS_OCI_AUTH for registry '${registry}'`, LogLevel.Trace);
+			const split = authContext.split('|');
+			const userToken = `${split[1]}:${split[2]}`;
+			return Buffer.from(userToken)
+				.toString('base64');
+		}
+	} else {
+		const homeDir = os.homedir();
+		if (homeDir) {
+			const dockerConfigPath = path.join(homeDir, '.docker', 'config.json');
+			if (await isLocalFile(dockerConfigPath)) {
+				try {
+					const dockerConfig: DockerConfigFile = JSON.parse((await readLocalFile(dockerConfigPath)).toString());
```

Comment:
Could use `jsonc.parse` for additional robustness. (Though Docker might actually be strict from what I have seen.)